### PR TITLE
calculatenetworkfee improvements

### PR DESCRIPTION
**Pull request by roman-khimov**

### CODE REVIEW

File: pkg/smartcontract/param_type.go [hunk 1]
```diff
@@ -139,6 +141,35 @@ func (pt *ParamType) DecodeBinary(r *io.BinReader) {
 	*pt = ParamType(r.ReadB())
 }
 
+// EncodeDefaultValue writes a script to push the default parameter value onto
+// the evaluation stack into the given writer. It's mostly useful for constructing
+// dummy invocation scripts when paramet types are known, but they can't be
```
Comment (AnnaShaleva) [inline pkg/smartcontract/param_type.go:146]: s/paramet/paarmeter

File: pkg/smartcontract/param_type.go [hunk 2]
```diff
@@ -139,6 +141,35 @@ func (pt *ParamType) DecodeBinary(r *io.BinReader) {
...
+// types the exact values can be arbitrarily long, but it tries to do something
+// reasonable in each case. For signatures, strings, arrays and "any" type a 64-byte
+// zero-filled value is used, hash160 and hash256 use appropriately sized values,
+// public key is represented by 33-byte value while 32 bytes are used for integer
+// and a simple push+convert is used for boolean. Other types produce no code at all.
+func (pt ParamType) EncodeDefaultValue(w *io.BinWriter) {
+	var b [64]byte
+
+	switch pt {
+	case AnyType, SignatureType, StringType, ByteArrayType:
```
Comment (AnnaShaleva) [inline pkg/smartcontract/param_type.go:157]: Are you sure we'd like to do this for Any, strings and byte arrays? It's pretty clear for signature, but other types are questionable.

File: pkg/services/rpcsrv/server.go [hunk 3]
```diff
@@ -758,26 +760,43 @@ func (s *Server) calculateNetworkFee(reqParams params.Params) (interface{}, *neo
...
+					paramz = make([]manifest.Parameter, nSigs)
+					for j := 0; j < nSigs; j++ {
+						paramz[j] = manifest.Parameter{Type: smartcontract.SignatureType}
+					}
+				}
+			}
+			inv := io.NewBufBinWriter()
+			for _, p := range paramz {
+				p.Type.EncodeDefaultValue(inv.BinWriter)
+			}
```
Comment (AnnaShaleva) [inline pkg/services/rpcsrv/server.go:791]: The code looks good, but do we have a compatibility test with C# node response for standard cases (sig/multisig witnesses)?

Comment (roman-khimov) [reply pkg/services/rpcsrv/server.go:791]: Our current tests cover all cases and transactions generated with `calculatenetworkfee` are pushed to the chain, so at least the fees calculated are sufficient. At the same time, we can add some tests to make sure they exactly match our expectations.

File: pkg/smartcontract/param_type.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/smartcontract/param_type.go:157]: Well, we either ignore them or try doing something like this. None of the approaches is perfect, but maybe the way it is now it'll handle a bit more cases (the most likely thing to be accepted by `verify` is some kind of signature).

File: pkg/core/blockchain.go [hunk 4]
```diff
@@ -2337,6 +2337,9 @@ func (bc *Blockchain) InitVerificationContext(ic *interop.Context, hash util.Uin
 func (bc *Blockchain) VerifyWitness(h util.Uint160, c hash.Hashable, w *transaction.Witness, gas int64) (int64, error) {
 	ic := bc.newInteropContext(trigger.Verification, bc.dao, nil, nil)
 	ic.Container = c
+	if tx, ok := c.(*transaction.Transaction); ok {
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/blockchain.go:2340]: Only transactions care about headers and transactions, everything else uses simple sig/multisig scripts.

File: docs/notary.md [hunk 5]
```diff
@@ -393,6 +376,9 @@ the steps to create a signature request:
      `Invocation` script either should push on stack signature bytes (one
      signature at max per one request) **or** (in case there's no ability to
      provide proper signature) **should be empty**.
+8. Calculate network fee for the transaction (that will be `NetworkFee`
+   transaction field). Use [func (*Client) CalculateNetworkFee](https://pkg.go.dev/github.com/nspcc-dev/neo-go@v0.99.2/pkg/rpcclient#Client.CalculateNetworkFee)
+   method with the main transaction given to it.
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply docs/notary.md:379]: OK, let's try it this way.

File: pkg/smartcontract/param_type.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/smartcontract/param_type.go]: Fixed.

File: pkg/services/rpcsrv/server.go (hunk 3 above)
Comment (roman-khimov) [reply pkg/services/rpcsrv/server.go]: Added some.

File: docs/notary.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply docs/notary.md:379]: Now looks better.

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**